tensorflow/compiler/xla/service/gpu/done_event_thunk.cc

```diff
@@ -0,0 +1,58 @@
+#include "tensorflow/compiler/xla/service/gpu/done_event_thunk.h"
+
+namespace xla {
+namespace gpu {
+
+using SpmdDoneEvents = absl::flat_hash_map<int, std::unique_ptr<se::Event>>;
+absl::flat_hash_map<int, SpmdDoneEvents*> all_done_events;
+
+void ResetEvents(size_t num_output) { all_done_events.clear(); }
+
+StatusOr<std::vector<std::unique_ptr<se::Event>>> TakeDoneEvents(
+    size_t output_index, int devices) {
+  auto it = all_done_events.find(output_index);
+  TF_RET_CHECK(it != all_done_events.end()) << "output index not found";
+  SpmdDoneEvents* output_done_events = it->second;
+  std::vector<std::unique_ptr<se::Event>> ret;
+  for (int i = 0;i < devices;++i) {
+    auto iter = output_done_events->find(i);
+    TF_RET_CHECK(iter != output_done_events->end()) << "done event not found";
+    ret.emplace_back(std::move(iter->second));
+    output_done_events->erase(iter);
+  }
+  all_done_events.erase(it);
+  return ret;
+}
+
+DoneEventThunk::DoneEventThunk(ThunkInfo thunk_info, size_t index)
+    : Thunk(Kind::kDoneEvent, thunk_info), output_index(index) {}
+
+Status DoneEventThunk::Initialize(const GpuExecutable& executable,
+                                  se::StreamExecutor* executor) {
+  if (executor->device_ordinal() == 0) {
+    all_done_events[output_index] = &done_events_;
+    done_events_.clear();
+  }
+  return Status::OK();
+}
+
+Status DoneEventThunk::ExecuteOnStream(const ExecuteParams& params) {
+  TF_ASSIGN_OR_RETURN(const GlobalDeviceId global_device_id,
+                      params.GetGlobalDeviceId());
+  TF_ASSIGN_OR_RETURN(const DeviceAssignment::LogicalID logical_id,
+                      params.device_assn->LogicalIdForDevice(global_device_id));
+  int PartitionId = logical_id.computation_id;
+  se::Stream& stream = *params.stream;
+  int device_ordinal = stream.parent()->device_ordinal();
+
+  auto done_event = absl::make_unique<se::Event>(stream.parent());
```

Comment:
Maybe this is a silly question: why `stream.parent()` is `se::Event`? I thought it is `GpuExecutor`.

Reply:
`se::Event` takes it as the input arg of its constructor function

Reply:
OK, get it. Here is a constructor but not a type_cast.